### PR TITLE
Update CRD API version to apiextensions.k8s.io/v1

### LINKED ISSUE #4969: CRDs should be upgraded from apiextensions.k8s.io/v1beta1 to apiextensions.k8s.io/v1
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug
/kind enhancement

**What happened**:
Building KubeVirt triggers the following warning a lot:
```
apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
```
We should make sure everything is upgraded to v1 in KubeVirt before k8s 1.22 comes out.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`):
- Kubernetes version (use `kubectl version`):
- VM or VMI specifications:
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by acardace**
Change apiextensions.k8s.io/v1beta1 to apiextensions.k8s.io/v1
as the former is deprecated and will be removed from Kubernetes 1.22.

Signed-off-by: Antonio Cardace <acardace@redhat.com>

**What this PR does / why we need it**:
Replaces all occurrences of `apiextensions.k8s.io/v1beta1` with `apiextensions.k8s.io/v1`
as the former is deprecated and will be removed from Kubernetes 1.22.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4969

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-operator/resource/generate/components/crds.go
Comment (xpivarc) [inline pkg/virt-operator/resource/generate/components/crds.go:74]: I think we should also update `&extv1beta1.CustomResourceDefinition`

**Comment by ichigo663** [comment]
@dhiller @xpivarc just keep this on hold, there's a lot more to migrate and I'm working on it.

File: pkg/virt-operator/resource/generate/components/crds.go
Comment (ashleyschuett) [inline pkg/virt-operator/resource/generate/components/crds.go:85]: I don't think we area always going to want to add all openapi spec values to every version. Since this is the version of the custom resource that we are serving. for example `virtualmachineinstance` if we decide to add or deprecate a field between versions that should be shown in the crd.

**Review by ashleyschuett — COMMENTED** [review]
I'm not sure why but the virtualmachinstance crd contains the virtualmachineinstancemigration schema, which is what's breaking the tests :/

File: pkg/virt-operator/resource/generate/components/crds.go
Comment (ichigo663) [reply pkg/virt-operator/resource/generate/components/crds.go:104]: Deleted it by mistake, will fix it. Thanks

File: pkg/virt-operator/resource/apply/reconcile.go
Comment (xpivarc) [inline pkg/virt-operator/resource/apply/reconcile.go:2]: Probably should be here.

File: tools/marketplace/helper/bundleHelper.go
Comment (xpivarc) [inline tools/marketplace/helper/bundleHelper.go:161]: Is this guaranteed?

**Review by xpivarc — CHANGES_REQUESTED** [review]
I had a first pass.

File: hack/builder/Dockerfile
Comment (acardace) [reply hack/builder/Dockerfile:95]: This is actually addressed in https://github.com/kubevirt/kubevirt/pull/5087, I will rebase once that gets merged but it's basically duplicated otherwise the tests would not pass (not completely true because the updated image is already in quay).

Anyway, we need to bump the operator-courier because v1.3.0 it does not support api-extensions v1 but just v1beta1.

File: pkg/virt-operator/resource/generate/install/strategy.go
Comment (acardace) [reply pkg/virt-operator/resource/generate/install/strategy.go:496]: It's deliberate, otherwise `manifests` will be aliased by the inner scope declaration, in this way I keep using the same var. I know this is not great...

File: pkg/virt-operator/resource/generate/install/strategy.go
Comment (acardace) [reply pkg/virt-operator/resource/generate/install/strategy.go:533]: In theory, yes, but it's actually used in `strategy_test.go` as well, idk I'd leave it there as the compiler most probably inlines it anyway(?).

File: tools/marketplace/helper/bundleHelper.go
Comment (acardace) [reply tools/marketplace/helper/bundleHelper.go:161]: Actually, you're right, it's not https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-priority.

I don't know why but I assumed it was like this, I'll think of something else.

File: pkg/virt-operator/resource/apply/reconcile.go
Comment (acardace) [reply pkg/virt-operator/resource/apply/reconcile.go]: oops, good catch, that's me trying (and failing) to exit vim.

File: pkg/virt-operator/resource/apply/crds.go
Comment (acardace) [reply pkg/virt-operator/resource/apply/crds.go]: Very true, updated the PR.

File: tools/marketplace/helper/bundleHelper.go
Comment (acardace) [reply tools/marketplace/helper/bundleHelper.go]: Updated.

File: pkg/virt-operator/resource/generate/install/strategy.go
Comment (xpivarc) [reply pkg/virt-operator/resource/generate/install/strategy.go:533]: The reason why I suggested it was to avoid bad usage in the future...

**Review by xpivarc — COMMENTED** [review]
There are some implications of this PR and they are worth noting.

1. We now require Kubernetes 1.17+ (we are ok here since we support 3version - currently 18,19,20)
2. Downgrade of Kubevirt will not work for Kubernetes 1.21+ (v1beta crd API will be removed)

/cc @davidvossel 
Can you have a look?

**Comment by acardace** [comment]
> There are some implications of this PR and they are worth noting.
> 
> 1. We now require Kubernetes 1.17+ (we are ok here since we support 3version - currently 18,19,20)
> 2. Downgrade of Kubevirt will not work for Kubernetes 1.21+ (v1beta crd API will be removed)
I think it's actually k8s 1.22+.

File: pkg/virt-operator/resource/apply/crds.go
Comment (acardace) [reply pkg/virt-operator/resource/apply/crds.go]: fixed, thank you.

File: pkg/virt-operator/resource/generate/install/strategy.go
Comment (acardace) [reply pkg/virt-operator/resource/generate/install/strategy.go]: done.

File: pkg/virt-operator/resource/generate/install/strategy.go
Comment (acardace) [reply pkg/virt-operator/resource/generate/install/strategy.go]: done.

File: pkg/virt-operator/resource/apply/crds_test.go
Comment (acardace) [reply pkg/virt-operator/resource/apply/crds_test.go]: Done, added 2 new unit tests in strategy_test.go which verifies that the ConfigMap gets correctly loaded whether encoded or not.

File: pkg/virt-operator/resource/apply/crds.go
Comment (xpivarc) [inline pkg/virt-operator/resource/apply/crds.go:112]: nit: Let's do `return` here.

**Review by xpivarc — APPROVED** [review]

File: pkg/virt-operator/resource/apply/crds.go
Comment (acardace) [reply pkg/virt-operator/resource/apply/crds.go:112]: if you did that you'd lose the logging message "crd %v is up-to-date" just below in case patching does not have any errors, I think that's pretty important.

Comment (xpivarc) [reply pkg/virt-operator/resource/apply/crds.go:112]: Fair enough :+1:

**Comment by xpivarc** [comment]
/lgtm
Thanks for the PR!

**Review by davidvossel — COMMENTED** [review]
/approve

This technique of encoding/decoding gzip+base64 should work given our current guarantees around updates where we require the kubevirt version to be less than or equal to the operator version that is deploying kubevirt.

**Comment by acardace** [comment]
@davidvossel I had to rebase and lgtm disappeared, I did no changes to the PR, can you please review it again? Thanks.

**Pull request merged**
